# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

[In this notebook you're provided with hints and brief instructions and thinking prompts. Don't ignore them as they are designed to equip you with the structure for the project and will help you analyze what you're doing on a deeper level. Before submitting your project, make sure you remove all hints and descriptions provided to you. Instead, make this report look as if you're sending it to your teammates to demonstrate your findings - they shouldn't know you had some external help from us! To help you out, we've placed the hints you should remove in square brackets.]

[Before you dive into analyzing your data, explain the purposes of the project and hypotheses you're going to test.]

## Open the data file and have a look at the general information. 

[Start with importing the libraries and loading the data. You may realise that you need additional libraries as you go, which is totally fine - just make sure to update this section when you do.]

In [1]:
# Loading all the libraries
import pandas as pd
import statistics

# Load the data
try:
    credit_scoring = pd.read_csv('credit_scoring_eng.csv')
except:
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv') 

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

[Now let's explore our data. You'll want to see how many columns and rows it has, look at a few rows to check for potential issues with the data.]

In [2]:
# Let's see how many rows and columns our dataset has
credit_scoring.shape


(21525, 12)

In [3]:
# let's print the first N rows
credit_scoring.head(20)
# negative days employed doesn't make sense.

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

[Describe what you see and notice in your printed data sample. Are there any issues that may need further investigation and changes?] 

Negative values in days employed and Capitalized Education

In [4]:
# Get info on data
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


[Are there missing values across all columns or just a few? Briefly describe what you see in 1-2 sentences.]

Missing values are only in days_employed and total_income

In [5]:
# Let's look at the filtered table with missing values in the the first column with missing data
credit_scoring[credit_scoring['days_employed'].isna()]
#missing values are only across days_employed and total_income

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

[Do missing values seem symmetric? Can we be sure in this assumption? Explain your thoughts briefly in this section. You may probably want to conduct further investigations, and count the missing values in all the rows with missing values to confirm the the missing samples are of the same size.]


In [6]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
credit_scoring_unemployed_NaN = credit_scoring[credit_scoring['days_employed'].isna()]
credit_scoring_unemployed_NaN = credit_scoring_unemployed_NaN[credit_scoring_unemployed_NaN['total_income'].isna()]
credit_scoring_unemployed_NaN.shape[0]
#credit_scoring.isna().sum() #confirms same amount missing in both days_employed and total_income

2174

**Intermediate conclusion**

[Does the number of rows in the filtered table match the number of missing values? What conclusion can we make from this?]

Yes. We can conclude that the missing values are the only ones rows that are missing.

[Calculate the percentage of the missing values compared to the whole dataset. Is it a considerably large piece of data? If so, you may want to fill the missing values. To do that, firstly we should consider whether the missing data could be due to the specific client characteristic, such as employment type or something else. You will need to decide which characteristic *you* think might be the reason. Secondly, we should check whether there's any dependence missing values have on the value of other indicators with the columns with identified specific client characteristic.]

[Explain your next steps and how they correlate with the conclusions you made so far.]



In [7]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
credit_scoring[credit_scoring['days_employed'].isna()]['days_employed']
#singular column, due to asking of singular column

12      NaN
26      NaN
29      NaN
41      NaN
55      NaN
         ..
21489   NaN
21495   NaN
21497   NaN
21502   NaN
21510   NaN
Name: days_employed, Length: 2174, dtype: float64

In [8]:
# Checking distribution
credit_scoring_unemployed_NaN_row_count = credit_scoring_unemployed_NaN.shape[0]
credit_scoring_employment = credit_scoring['days_employed'].shape[0]
credit_scoring_distribution = 'Missing value distribution to the column is ' + str(credit_scoring_unemployed_NaN_row_count/credit_scoring_employment * 100) + '%' 
credit_scoring_distribution

'Missing value distribution to the column is 10.099883855981417%'

[Describe your findings here.]

Missing value distribution to the column is 10.099883855981417%

[Propose your ideas on why you think the values might be missing. Do you think they are missing randomly or there are any patterns?]

Human error. They probably put it in wrong.

[Let's start checking whether the missing values are random.]

In [9]:
# Checking the distribution in the whole dataset
#using all NaN elements divided by total size of table 
credit_scoring_NaN_total = credit_scoring_unemployed_NaN_row_count + credit_scoring[credit_scoring['total_income'].isna()].shape[0]
credit_total_size = credit_scoring.size 
credit_distribution = 'Missing value distribution to the whole data set is ' + str(credit_scoring_NaN_total/credit_total_size * 100 ) + '%' 
credit_distribution

# Has to be Human error. There's no way that non retirees other than interns cannot have an income, or attempt to borrow.
#The overall percentage compared to the rest of the table is too insignifigant to make a difference.
#(2174 + 2174)/credit_total_size * 100 

'Missing value distribution to the whole data set is 1.683313975996903%'

**Intermediate conclusion**

[Is the distribution in the original dataset similar to the distribution of the filtered table? What does that mean for us?]

[If you think we can't make any conclusions yet, let's investigate our dataset further. Let's think about other reasons that could lead to data missing and check if we can find any patterns that may lead us to thinking that the missing values are not random. Because this is your work, this is section is optional.]

In [10]:
# Check for other reasons and patterns that could lead to missing values
#NaN values seemingly look mostly like mistakes in entering values 

**Intermediate conclusion**

[Can we finally confirm that missing values are accidental? Check for anything else that you think might be important here.]

In [11]:
# Checking for other patterns - explain which


**Conclusions**

[Did you find any patterns? How did you come to this conclusion?]

[Explain how you will address the missing values. Consider the categories in which values are missing.]

[Briefly plan your next steps for transforming data. You will probably need to address different types of issues: duplicates, different registers, incorrect artifacts, and missing values.]

## Data transformation

[Let's go through each column to see what issues we may have in them.]

[Begin with removing duplicates and fixing educational information if required.]

In [12]:
# Let's see all values in education column to check if and what spellings will need to be fixed
credit_scoring['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [13]:
# Fix the registers if required
credit_scoring['education'] = credit_scoring['education'].str.lower()


In [14]:
# Checking all the values in the column to make sure we fixed them
credit_scoring['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

[Check the data the `children` column]

In [15]:
# Let's see the distribution of values in the `children` column
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

[Are there any strange things in the column? If yes, how high is the percentage of problematic data? How could they have occurred? Make a decision on what you will do with this data and explain you reasoning.]

In [16]:
# [fix the data based on your decision]
#families with 20 children or -1 children seem like huge outliers -1 children seems impossible.
#If your data has significant outliers, calculate the median using the median() method. https://practicum.yandex.com/trainer/data-scientist/lesson/a5a59234-272d-4bad-a9ca-8d87dba9a1f1/
#I'll calculate the median without the outliers. Adding importing statistics
children_total = credit_scoring['children'].shape[0] 
problematic_data = 76 + 41
percentage = str(problematic_data/children_total * 100)
print('total percentage of problematic data is ' + percentage + '%') #forgot to calculate the problematic data percentage last time

from statistics import median
children_median = int(median([0,1,2,3,4,5])) # = without outliers
credit_scoring = credit_scoring.replace({'children':{-1:children_median,20:children_median}}) #hashes in hashes

total percentage of problematic data is 0.5435540069686411%


In [17]:
# Checking the `children` column again to make sure it's all fixed

credit_scoring['children'].value_counts()

0    14149
1     4818
2     2178
3      330
4       41
5        9
Name: children, dtype: int64

[Check the data in the `days_employed` column. Firstly think about what kind of issues could there be and what you may want to check and how you will do it.]

In [18]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage

#all_NaN = credit_scoring[credit_scoring['days_employed'].isna()]['days_employed'].shape[0] #2174 we don't need NaN right now
all_neg = credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].shape[0] #15906
total_days = credit_scoring['days_employed'].shape[0] #21525

problematic_data_days = "Total percentage is " + str(all_neg/total_days * 100) +"%"
problematic_data_days

'Total percentage is 73.89547038327527%'

[If the amount of problematic data is high, it could've been due to some technical issues. We may probably want to propose the most obvious reason why it could've happened and what the correct data might've been, as we can't drop these problematic rows.]

In [19]:
# Address the problematic values, if they exist
#All negatives are probably one person changing the values negative by accident.
#https://pastebin.com/P8N7k2gv

#def to_positive(x):
 #   if x < 0:
  #      return -x
    
#simpler than .replace() each data 73% since the data is so big and logical_indexing and .loc wont work 
    
credit_scoring['days_employed'] = credit_scoring['days_employed'].abs()    #absolute value is distance from zero.

In [20]:
# Check the result - make sure it's fixed
credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].shape[0]

0

[Let's now look at the client's age and whether there are any issues there. Again, think about what can data can be strange in this column, i.e. what cannot be someone's age.]

In [21]:
# Check the `dob_years` for suspicious values and count the percentage
credit_scoring['dob_years'].value_counts()


35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

[Decide what you'll do with the problematic values and explain why.]

In [22]:
# Address the issues in the `dob_years` column, if they exist
#Nobody works at 0 years old. Nobody.
credit_scoring['dob_years'].median() #median because outliers
credit_scoring['dob_years'] = credit_scoring['dob_years'].replace(0,42)

In [23]:
# Check the result - make sure it's fixed
credit_scoring[credit_scoring['dob_years'] == 0].shape[0]

0

[Now let's check the `family_status` column. See what kind of values there are and what problems you may need to address.]

In [24]:
# Let's see the values for the column

credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [25]:
# Address the problematic values in `family_status`, if they exist
#none here

In [26]:
# Check the result - make sure it's fixed
credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

[Now let's check the `gender` column. See what kind of values there are and what problems you may need to address]

In [27]:
# Let's see the values in the column
credit_scoring['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [28]:
# Address the problematic values, if they exist
#XNA could be due to third gender
#https://bhadreshpsavani.medium.com/understanding-data-before-data-preprocessing-using-python-data-science-tutorial-series-5499961f3aaf 
def gender_number(value): #We need this because we can't get means from strings.
    if value == 'F':
        return 1
    if value == 'M':
        return 2
    if value == 'XNA':
        return 3

credit_scoring['gender_number'] = credit_scoring['gender'].apply(gender_number) 

gender_median = int(credit_scoring['gender_number'].mean())
if gender_median == 1:
    gender_median = 'F'
else:
    gender_median = 'M'
print(gender_median)
credit_scoring = credit_scoring.replace({'gender':{'XNA':gender_median}}) #if it is less than 5%, get rid of it. Like you said.
credit_scoring = credit_scoring.drop(columns=['gender_number'])

F


In [29]:
# Check the result - make sure it's fixed

credit_scoring['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

[Now let's check the `income_type` column. See what kind of values there are and what problems you may need to address]

In [30]:
# Let's see the values in the column
credit_scoring['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [31]:
# Address the problematic values, if they exist

In [32]:
# Check the result - make sure it's fixed

credit_scoring['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

[Now let's see if we have any duplicates in our data. If we do, you'll need to decide what you will do with them and explain why.]

In [33]:
# Checking duplicates
credit_scoring.duplicated().value_counts()


False    21453
True        72
dtype: int64

In [34]:
# Address the duplicates, if they exist
credit_scoring[credit_scoring.duplicated()]

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[72 rows x 12 columns]

In [35]:
# Last check whether we have any duplicates
credit_scoring = credit_scoring.drop_duplicates()

In [36]:
# Check the size of the dataset that you now have after your first manipulations with it
credit_scoring.size


257436

[Describe your new dataset: briefly say what's changed and what's the percentage of the changes, if there were any.]

All the duplicates were removed. No duplicates, Difference capitalizations Education, no Negative incomes


# Working with missing values

[To speed up working with some data, you may want to work with dictionaries for some values, where IDs are provided. Explain why and which dictionaries you will work with.]

In [37]:
# Find the dictionaries
import numpy as np

### Restoring missing values in `total_income`

[Briefly state which column(s) have values missing that you need to address. Explain how you will fix them.]


[Start with addressing total income missing values. Create and age category for clients. Create a new column with the age category. This strategy can help with calculating values for the total income.]


In [38]:
# Let's write a function that calculates the age category

#dob_years - client's age in years
#credit_scoring['dob_years']

def age(dob_years):
    return dob_years

In [39]:
# Test if the function works
credit_scoring['dob_years'].apply(age)

0        42
1        36
2        33
3        32
4        53
         ..
21520    43
21521    67
21522    38
21523    38
21524    40
Name: dob_years, Length: 21453, dtype: int64

In [40]:
# Creating new column based on function

credit_scoring['age'] = credit_scoring['dob_years'].apply(age)

In [41]:
# Checking how values in the new column

credit_scoring['age'].value_counts()

42    696
35    616
40    607
41    605
34    601
38    597
33    581
39    572
31    559
36    554
44    545
29    544
30    537
37    536
48    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
55    443
59    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: age, dtype: int64

[Think about the factors on which income usually depends. Eventually, you will want to find out whether you should use mean or median values for replacing missing values. To make this decision you will probably want to look at the distribution of the factors you identified as impacting one's income.]

[Create a table that only has data without missing values. This data will be used to restore the missing values.]

In [42]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
credit_scoring #thanks for that. I didn't realize that was needed. 

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  age  
0         40620.102    purchase of the house   42  
1         17932.802             car purchase   36  
2         23341.752    purchase of the house   33  
3         42820.568  supplementary education   32  
4         25378.572        to have a wedding   53  
...             ...                      ...  ...  
21520     35966.698     housing transactions   43  
21521     24959.969        purchase of a car   67  
21522     14347.610                 property   38  
21523     39054.888        buying my own car   38  
21524     13127.587             to buy a car   40  

[21453 rows x 13 columns]

In [43]:
# Look at the mean values for income based on your identified factors
income_mean = credit_scoring['total_income'].mean()

In [44]:
# Look at the median values for income based on your identified factors
income_median = credit_scoring['total_income'].median()

[Repeat such comparisons for multiple factors. Make sure you consider different aspects and explain your thinking process.]



[Make a decision on what characteristics define income most and whether you will use a median or a mean. Explain why you made this decision]

No huge outliers and no discernable patern between this and other columns.


In [45]:
#  Write a function that we will use for filling in missing values
#no signifigant outliers so we can use mean.
#https://towardsdatascience.com/5-methods-to-check-for-nan-values-in-in-python-3f21ddd17eed
#originally was going to use .mean()

def NaN_fill(value):
    if value != value:
        value = income_mean
        return value
    return value

In [46]:
# Check if it works
credit_scoring['total_income'].apply(NaN_fill)

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21453, dtype: float64

In [47]:
# Apply it to every row
credit_scoring['total_income'] = credit_scoring['total_income'].apply(NaN_fill)

In [48]:
# Check if we got any errors
credit_scoring[credit_scoring['total_income'].isna()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, age]
Index: []

[If you've came across errors in preparing the values for missing data, it probably means there's something special about the data for the category. Give it some thought - you may want to fix some things manually, if there's enough data to find medians/means.]


In [49]:
# Replacing missing values if there are any errors
#None so far

[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [50]:
# Checking the number of entries in the columns
credit_scoring.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21453 entries, 0 to 21524
Data columns (total 13 columns):
children            21453 non-null int64
days_employed       19351 non-null float64
dob_years           21453 non-null int64
education           21453 non-null object
education_id        21453 non-null int64
family_status       21453 non-null object
family_status_id    21453 non-null int64
gender              21453 non-null object
income_type         21453 non-null object
debt                21453 non-null int64
total_income        21453 non-null float64
purpose             21453 non-null object
age                 21453 non-null int64
dtypes: float64(2), int64(6), object(5)
memory usage: 2.3+ MB


###  Restoring values in `days_employed`

[Think about the parameters that may help you restore the missing values in this column. Eventually, you will want to find out whether you should use mean or median values for replacing missing values. You will probably conduct a research similar to the one you've done when restoring data in a previous column.]

In [51]:
# Distribution of `days_employed` medians based on your identified parameters
days_employed_median_group = credit_scoring[credit_scoring['income_type']!='retiree'] #retiree's aren't employed
days_employed_median = days_employed_median_group[credit_scoring['days_employed'].notna()]['days_employed'].median()
days_employed_total = days_employed_median_group[days_employed_median_group['days_employed'].notna()]['days_employed'].sum()
str(days_employed_median/days_employed_total * 100) + '%'


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


'0.004271962258683324%'

In [52]:
# Distribution of `days_employed` means based on your identified parameters
days_employed_mean = days_employed_median_group[days_employed_median_group['days_employed'].notna()]['days_employed'].mean()
str(days_employed_mean/days_employed_total * 100) + '%'

'0.006286145335680161%'

[Decide what you will use: means or medians. Explain why.]

Mean, because there is no discernable distinguishable outliers.

In [53]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
#https://pastebin.com/ZVXYcmyU mistakes and notes
def mean_maker(value):
    if value != value: 
        value = credit_scoring['days_employed'].shape[0]/credit_scoring['days_employed'].sum() #IT WORKS, didn't think it would work.
    return value
    

In [54]:
# Check that the function works

days_employed_median_group['days_employed'] = days_employed_median_group['days_employed'].apply(mean_maker)
days_employed_median_group[days_employed_median_group['days_employed'].isna()].shape[0]

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0

In [55]:
# Apply function to the income_type
credit_scoring['income_type'] = credit_scoring['income_type'].apply(mean_maker)

In [56]:
# Check if function worked
credit_scoring[credit_scoring['income_type'].isna()].shape[0] #strings don't have means!

0

In [57]:
# Replacing missing values
#https://pastebin.com/ERe3U8cR mistakes were made
#https://stackoverflow.com/questions/45803676/python-pandas-loc-filter-for-list-of-values
all_retirees = list(credit_scoring[credit_scoring['income_type']=='retiree']['days_employed'].index.values)
credit_scoring['days_employed'] = credit_scoring['days_employed'].apply(mean_maker)
credit_scoring.loc[all_retirees,'days_employed'] = 0

[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [58]:
# Check the entries in all columns - make sure we fixed all missing values
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21453 entries, 0 to 21524
Data columns (total 13 columns):
children            21453 non-null int64
days_employed       21453 non-null float64
dob_years           21453 non-null int64
education           21453 non-null object
education_id        21453 non-null int64
family_status       21453 non-null object
family_status_id    21453 non-null int64
gender              21453 non-null object
income_type         21453 non-null object
debt                21453 non-null int64
total_income        21453 non-null float64
purpose             21453 non-null object
age                 21453 non-null int64
dtypes: float64(2), int64(6), object(5)
memory usage: 2.9+ MB


## Categorization of data

[To answer the questions and test the hypotheses, you will want to work with categorized data. Look at the questions that were posed to you and that you should answer. Think about which of the data will need to be categorized to answer these questions. Below you will find a template through which you can work your way when categorizing data. The first step-by-step processing covers the text data; the second one addresses the numerical data that needs to be categorized. You can use both or none of the suggested instructions - it's up to you.]

[Despite of how you decide to address the categorization, make sure to provide clear explanation of why you made your decision. Remember: this is your work and you make all decisions in it.]


In [59]:
# Print the values for your selected data for categorization

print(credit_scoring['children'])
print(credit_scoring['family_status'])
print(credit_scoring['total_income'])
print(credit_scoring['purpose'])
print(credit_scoring['debt']) 

0        1
1        1
2        0
3        3
4        0
        ..
21520    1
21521    0
21522    1
21523    3
21524    2
Name: children, Length: 21453, dtype: int64
0                  married
1                  married
2                  married
3                  married
4        civil partnership
               ...        
21520    civil partnership
21521              married
21522    civil partnership
21523              married
21524              married
Name: family_status, Length: 21453, dtype: object
0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21453, dtype: float64
0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21520       housing transaction

[Let's check unique values]

In [60]:
# Check the unique values
print(credit_scoring['children'].value_counts())
print(credit_scoring['family_status'].value_counts())
print(credit_scoring['total_income'].value_counts())
print(credit_scoring['purpose'].value_counts()) #realized purpose is the only one uncategorized/simplified
print(credit_scoring['debt'].value_counts()) #assuming debt > 0 means they don't pay loans on time

0    14090
1     4808
2     2175
3      330
4       41
5        9
Name: children, dtype: int64
married              12339
civil partnership     4150
unmarried             2810
divorced              1195
widow / widower        959
Name: family_status, dtype: int64
26787.568355    2102
17312.717000       2
31791.384000       2
42413.096000       2
20226.282000       1
                ... 
48796.341000       1
34774.610000       1
15710.698000       1
19232.334000       1
9591.824000        1
Name: total_income, Length: 19349, dtype: int64
wedding ceremony                            791
having a wedding                            767
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house    

[What main groups can you identify based on the unique values?]

Education/Real Estate/Car/wedding

[Based on these themes, we will probably want to categorize our data.]



In [61]:
# Let's write a function to categorize the data based on common topics

def purpose_sort(purpose):
    if 'educat' in purpose:
        return 'education'
    if 'property' in purpose or 'real estate' in purpose or 'hous' in purpose:
        return 'real estate'
    if 'wedding' in purpose:
        return 'wedding'
    if 'car' in purpose:
        return 'car'
    if purpose == 'civil partnership' or purpose == 'married':
        return 'married'
    if 'unmarried' == purpose or 'widow' in purpose or 'divorced' in purpose:
        return 'single'

In [62]:
# Create a column with the categories and count the values for them
credit_scoring['general_purpose'] = credit_scoring['purpose'].apply(purpose_sort)
print(credit_scoring['general_purpose'].value_counts())
credit_scoring['relationship_status'] = credit_scoring['family_status'].apply(purpose_sort)
print(credit_scoring['relationship_status'].value_counts())

real estate    10811
car             4306
education       3517
wedding         2323
Name: general_purpose, dtype: int64
married    16489
single      4964
Name: relationship_status, dtype: int64


[If you decide to categorize the numerical data, you'll need to come up with the categories for it too.]

In [63]:
# Looking through all the numerical data in your selected column for categorization
print(credit_scoring['total_income'])

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21453, dtype: float64


In [64]:
# Getting summary statistics for the column
credit_scoring['total_income'].describe()

count     21453.000000
mean      26787.568355
std       15647.461536
min        3306.762000
25%       17219.352000
50%       24966.693000
75%       31331.009000
max      362496.645000
Name: total_income, dtype: float64

[Decide what ranges you will use for grouping and explain why.]

https://money.usnews.com/money/personal-finance/family-finance/articles/where-do-i-fall-in-the-american-economic-class-system

Poor or near-poor	$32,048 or less

Lower-middle class	$32,048 - $53,413

Middle class	$53,413 - $106,827

Upper-middle class	$106,827 - $373,894

Rich	$373,894 and up

An average estimate, based on a finance website. They know finance better than me. A wise man once said, we are closer to being homeless than being bezos or musk.

In [65]:
# Creating function for categorizing into different numerical groups based on ranges

def income_class(value):
    if value < 32048:
        return 'Poor'
    if 32048 <= value < 53413:
        return 'Lower-Middle Class'
    if 53413 <= value < 106827:
        return 'Middle Class'
    if 106827 <= value < 373894:
        return 'Upper Middle Class'
    if 373894 < value:
        return 'Rich'


In [66]:
# Creating column with categories
credit_scoring['income_class'] = credit_scoring['total_income'].apply(income_class)

In [67]:
# Count each categories values to see the distribution
credit_scoring['income_class'].value_counts()

Poor                  16408
Lower-Middle Class     3999
Middle Class            969
Upper Middle Class       77
Name: income_class, dtype: int64

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [68]:
# Check the children data and paying back on time
#We are assuming people who pay back on time have no debt. 
#https://pastebin.com/N1cSXFEN mistakes were made, we only need people with no debt.
#https://pastebin.com/0cSJ4QcT refined a second time.
#https://pastebin.com/H2SyG56h KEYS ARE EASIER THAN RANGE

no_debt = credit_scoring[credit_scoring['debt']==0]

# Calculating default-rate based on the number of children
no_debt_children_dictionary = no_debt['children'].value_counts().to_dict() #we need to access the value as key value pairs

def children_rate_maker(dictionary, child_amount):
        total = credit_scoring[credit_scoring['children'] == child_amount].shape[0] #all children of an amount
        value = dictionary[child_amount] #key value pair accessed by amount
        rate =  value/total * 100 #percentage
        if child_amount == 1:
            print('people with ' + str(child_amount) + ' child have a rate of ' + str(rate) + '% of paying back on time')
        print('people with ' + str(child_amount) + ' children have a rate of ' + str(rate) + '% of paying back on time')
        
for i in no_debt_children_dictionary.keys(): #I'm not calling a function multiple times. So I'll loop it for me
    children_rate_maker(no_debt_children_dictionary, i)

people with 0 children have a rate of 92.4556422995032% of paying back on time
people with 1 child have a rate of 90.76539101497504% of paying back on time
people with 1 children have a rate of 90.76539101497504% of paying back on time
people with 2 children have a rate of 90.66666666666666% of paying back on time
people with 3 children have a rate of 91.81818181818183% of paying back on time
people with 4 children have a rate of 90.2439024390244% of paying back on time
people with 5 children have a rate of 100.0% of paying back on time


**Conclusion**

[Write your conclusions based on your manipulations and observations.]

People with alot or no children tend to pay back on time more. My speculation is familys with 5 children can't afford debt (or can afford 5 children) so they pay off on time and 0 children spend less on children to pay it off.

**Is there a correlation between family status and paying back on time?**

In [69]:
# Check the family status data and paying back on time
#family status is now relationship_status, simplified.
#https://pastebin.com/fpygZqDS

no_debt_family_dictionary = no_debt['relationship_status'].value_counts().to_dict()


# Calculating default-rate based on family status

def status_rate_maker(dictionary,relationship_status):
        total = credit_scoring[credit_scoring['relationship_status'] == relationship_status].shape[0]
        value = dictionary[relationship_status]
        rate =  value/total * 100
        print(str(relationship_status) + ' people have a rate of ' + str(rate) + '% of paying back on time')

for i in no_debt_family_dictionary.keys():
    status_rate_maker(no_debt_family_dictionary,i)



married people have a rate of 92.0007277578992% of paying back on time
single people have a rate of 91.49879129734086% of paying back on time


**Conclusion**

[Write your conclusions based on your manipulations and observations.]

**Is there a correlation between income level and paying back on time?**

In [70]:
# Check the income level data and paying back on time
#income has been sorted into income_class

income_class_dictionary = no_debt['income_class'].value_counts().to_dict()

# Calculating default-rate based on income level

def class_rate_maker(dictionary,income_class):
        total = credit_scoring[credit_scoring['income_class'] == income_class].shape[0]
        value = dictionary[income_class]
        rate =  value/total * 100
        print(str(income_class) + ' people have a rate of ' + str(rate) + '% of paying back on time')

for i in income_class_dictionary.keys():
    class_rate_maker(income_class_dictionary,i)




Poor people have a rate of 91.55899561189663% of paying back on time
Lower-Middle Class people have a rate of 92.92323080770193% of paying back on time
Middle Class people have a rate of 92.98245614035088% of paying back on time
Upper Middle Class people have a rate of 93.5064935064935% of paying back on time


**Conclusion**

[Write your conclusions based on your manipulations and observations.]

More income, higher chance of paying back on time.

**How does credit purpose affect the default rate?**

In [71]:
# Check the percentages for default rate for each credit purpose and analyze them
#using general purpose


purpose_dictionary = no_debt['general_purpose'].value_counts().to_dict()


def purpose_rate_maker(dictionary,general_purpose):
        total = credit_scoring[credit_scoring['general_purpose'] == general_purpose].shape[0]
        value = dictionary[general_purpose]
        rate =  value/total * 100
        print('People who take out loans for ' + str(general_purpose) + ' have a rate of ' + str(rate) + '% of paying back on time')

for i in purpose_dictionary.keys():
    purpose_rate_maker(purpose_dictionary,i)



People who take out loans for real estate have a rate of 92.76662658403478% of paying back on time
People who take out loans for car have a rate of 90.64096609382257% of paying back on time
People who take out loans for education have a rate of 90.7023030992323% of paying back on time
People who take out loans for wedding have a rate of 91.99311235471373% of paying back on time


**Conclusion**

[Write your conclusions based on your manipulations and observations.]

Weddings and Houses have a higher rate of paying back on time than cars and education. My guess is due to the finality of a house (and the risk of getting evicted) and wedding (to have and to hold till death do us part) vs college debt and constant car payments and repairs and gas

# General Conclusion 

[List your conclusions in this final section. Make sure you include all your important conclusions you made that led you to the way you processed and analyzed the data. Cover the missing values, duplicates, and possible reasons and solutions for problematic artifacts that you had to address.]

All the duplicates were removed. No duplicates, Difference capitalizations in Education, no Negative incomes. Family Status and Income and Purpose were simplified.


[List your conclusions regarding the posed questions here as well.]

My conclusion, higher income, paying for things all at once like a wedding or morgage, and either having no extra expenses like children or having many expenses, like too much children show fiscal responsibility on what they can handle.Thus they can pay back on time.
